### PR TITLE
Adding tensor shape expansion/cleanup passes.

### LINKED ISSUE #8394: Propagate dynamic dimensions across calls.
I swear we do this somewhere but I can't find it.

```mlir
#map0 = affine_map<(d0, d1) -> (d0, d1)>
#map1 = affine_map<(d0, d1) -> (d0)>
module {
  func @bar(%arg0: !hal.buffer_view) attributes {iree.abi.stub} {
    %0 = hal.buffer_view.dim<%arg0 : !hal.buffer_view>[0] : index
    %1 = hal.buffer_view.dim<%arg0 : !hal.buffer_view>[1] : index
    %2 = hal.tensor.import %arg0 : !hal.buffer_view -> tensor<?x?xi8>{%0, %1}
    call @_bar(%2) : (tensor<?x?xi8>) -> ()
    return
  }
  flow.executable private @_bar_dispatch_0 {
    flow.dispatch.entry public @_bar_dispatch_0 attributes {workgroup_rank = 3 : index}
    builtin.module {
      func @_bar_dispatch_0(%arg0: !flow.dispatch.tensor<readonly:?x?xi8>, %arg1: index, %arg2: index, %arg3: index, %arg4: !flow.dispatch.tensor<writeonly:?xi32>) {
        %c0_i32 = arith.constant 0 : i32
        %0 = flow.dispatch.tie_shape %arg0 : !flow.dispatch.tensor<readonly:?x?xi8>{%arg1, %arg2}
        %1 = flow.dispatch.tie_shape %arg4 : !flow.dispatch.tensor<writeonly:?xi32>{%arg3}
        %2 = flow.dispatch.tensor.load %0, offsets = [0, 0], sizes = [%arg1, %arg2], strides = [1, 1] : !flow.dispatch.tensor<readonly:?x?xi8>{%arg1, %arg2} -> tensor<?x?xi8>
        %3 = linalg.init_tensor [%arg3] : tensor<?xi32>
        %4 = linalg.fill(%c0_i32, %3) : i32, tensor<?xi32> -> tensor<?xi32> 
        %5 = linalg.generic {indexing_maps = [#map0, #map1], iterator_types = ["parallel", "reduction"]} ins(%2 : tensor<?x?xi8>) outs(%4 : tensor<?xi32>) {
        ^bb0(%arg5: i8, %arg6: i32):
          %6 = arith.extsi %arg5 : i8 to i32
          %7 = arith.addi %6, %arg6 : i32
          linalg.yield %7 : i32
        } -> tensor<?xi32>
        flow.dispatch.tensor.store %5, %1, offsets = [0], sizes = [%arg3], strides = [1] : tensor<?xi32> -> !flow.dispatch.tensor<writeonly:?xi32>{%arg3}
        return
      }
    }
  }
  func private @_bar(%arg0: tensor<?x?xi8>) {
    %c1 = arith.constant 1 : index
    %c0 = arith.constant 0 : index
    %0 = tensor.dim %arg0, %c0 : tensor<?x?xi8>
    %1 = tensor.dim %arg0, %c1 : tensor<?x?xi8>
    %2 = flow.dispatch @_bar_dispatch_0::@_bar_dispatch_0[%0, %c1, %c1](%arg0, %0, %1, %0) : (tensor<?x?xi8>{%0, %1}, index, index, index) -> tensor<?xi32>{%0}
    check.expect_eq(%2, %2) : tensor<?xi32>
    return
  }
}
```

Could use DFX to perform the analysis for dim resolution but I think a simple expansion and an IPO pass (whenever that gets implemented) would handle all the same cases. Doing this in flow by expanding args/results and inserting flow.tensor.reshape ops on either side should help things fold away (as we resolve tensor.dim through any shape-carrying flow op). Doing this earlier in the pipeline would be useful as it may aid linalg/fusion (knowing which dims are equivalent, etc) so long as the reshapes and dims folded and didn't cause trouble.

### PR DESCRIPTION
**Pull request by benvanik**
This expands tensors to include their dynamic dimensions as SSA
values across global store/load, call/return, and br/cond_br.
This allows us to resolve the dimensions to SSA values without
any analysis; things like linalg fusion can know for certain how
shape dimensions equate to each other and simple localized patterns
can now effectively IPO shapes.

This brought back flow.tensor.tie_shape so that we can compose
all the optimizations we have in the Util/ transforms for globals
and branch/func propagation/collapsing/etc. Sucks, but may actually
be the right design choice for this level of IR.

As part of the cleanup it's asserted that no more tensor.dim ops
remain in the IR: at the end of the flow dialect there should be
no need for them.

Fixes #8394.

### CODE REVIEW

File: iree/compiler/Dialect/Flow/Transforms/CleanupTensorShapes.cpp [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    getOperation()->walk(
+        [&](Operation *op) {
+          if (auto tieShapeOp = dyn_cast<IREE::Flow::TensorTieShapeOp>(op)) {
+            tieShapeOp.replaceAllUsesWith(tieShapeOp.operand());
+            tieShapeOp.erase();
+          } else if (isa<tensor::DimOp>(op) || isa<tensor::RankOp>(op)) {
+            op->emitOpError()
+                << "unexpected during shape cleanup; dynamic dimensions must "
+                   "have been resolved prior to leaving the flow dialect";
+            foundBadOps = true;
```
Comment (ScottTodd) [inline iree/compiler/Dialect/Flow/Transforms/CleanupTensorShapes.cpp:34]: Want to add a lit test for this pass?

File: iree/compiler/Dialect/Flow/Transforms/test/expand_tensor_shapes.mlir [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  // CHECK-NEXT: %[[SEL_D1:.+]] = arith.select %[[COND]], %[[ARG0_D1]], %[[ARG1_D1]]
+  // CHECK-NEXT: %[[SEL_D2:.+]] = arith.select %[[COND]], %[[ARG0_D2]], %[[ARG1_D2]]
+  // CHECK-NEXT: %[[SEL_TIED:.+]] = flow.tensor.tie_shape %[[SEL_TENSOR]] : tensor<4x?x?x2xf32>{%[[SEL_D1]], %[[SEL_D2]]}
+  %0 = arith.select %cond, %arg0, %arg1 : tensor<4x?x?x2xf32>
+
+  // CHECK-NEXT: util.do_not_optimize(%[[SEL_TIED]])
+  util.do_not_optimize(%0) : tensor<4x?x?x2xf32>
+
+  return
+}
```
Comment (ScottTodd) [inline iree/compiler/Dialect/Flow/Transforms/test/expand_tensor_shapes.mlir:155]: fyi there is no newline character at the end of this file (check your editor settings?)

File: iree/compiler/Dialect/Flow/Transforms/Passes.cpp [hunk 3]
```diff
@@ -116,59 +116,63 @@ void buildFlowTransformPassPipeline(OpPassManager &passManager,
...
       // Input should now be legal.
-      .addPass(createVerifyInputLegalityPass);
+      .addPass(IREE::Flow::createVerifyInputLegalityPass);
 
   passManager.addPass(mlir::createLinalgNamedOpConversionPass());
+
+  // Expand tensor shapes into SSA values and optimize the whole program.
+  // The more we are able to equate shape dimensions at this level the better
+  // our fusions will be.
+  passManager.addPass(IREE::Flow::createExpandTensorShapesPass());
```
Comment (ScottTodd) [inline iree/compiler/Dialect/Flow/Transforms/Passes.cpp:138]: Is there any integration / e2e test that this enable support for, or it this just an optimization? Should we have [more] programs in our benchmark suite that include dynamic shapes?

**Review by ScottTodd — APPROVED** [review]

**Comment by bjacob** [comment]
Thanks fixes my issue / unblocks dynamic-shapes quantized_matmul e2e tests. Please merge!

File: iree/compiler/Dialect/Flow/Transforms/Passes.cpp (hunk 3 above)
Comment (bjacob) [reply iree/compiler/Dialect/Flow/Transforms/Passes.cpp:138]: Yes, this enables me to write e2e tests that I can't write so far, specifically about any e2e test involving functions taking a dynamic-shapes tensor as argument and performing a `tensor.dim` on it.

Comment (bjacob) [reply iree/compiler/Dialect/Flow/Transforms/Passes.cpp:138]: Here's a e2e test that could be added to `iree/test/e2e/regression` (failing without this PR, succeeding with this PR):

```
func private @foo(%lhs : tensor<?x?xi8>) -> tensor<?xi32> {
  %c_0_index = arith.constant 0 : index
  %c_1_index = arith.constant 1 : index
  %m_size = tensor.dim %lhs, %c_0_index : tensor<?x?xi8>
  %c_0 = arith.constant 0 : i32

  // compute the sums along rows of %lhs.
  %lhs_i32 = arith.extsi %lhs : tensor<?x?xi8> to tensor<?x?xi32>
  %init_lhs_sums_uninitialized = linalg.init_tensor [%m_size] : tensor<?xi32>
  %zero_lhs_sums = linalg.fill(%c_0, %init_lhs_sums_uninitialized) : i32, tensor<?xi32> -> tensor<?xi32>
  return %zero_lhs_sums : tensor<?xi32>
}

func @bar(%lhs : tensor<?x?xi8>) {
    %a = call @foo(%lhs) : (tensor<?x?xi8>) -> tensor<?xi32>
    check.expect_eq(%a, %a) : tensor<?xi32>
    return
}
```

Comment (bjacob) [reply iree/compiler/Dialect/Flow/Transforms/Passes.cpp:138]: I tried hard to provide a minimized testcase, but the best I came up with triggers a null deref (with this PR applied):

```
func private @reduce(%matrix : tensor<?x?xi8>) -> tensor<?xi32> {
  %c_0_index = arith.constant 0 : index
  %c_1_index = arith.constant 1 : index
  %m_size = tensor.dim %matrix, %c_0_index : tensor<?x?xi8>
  %c_0 = arith.constant 0 : i32

  %matrix_i32 = arith.extsi %matrix : tensor<?x?xi8> to tensor<?x?xi32>
  %init_matrix_uninitialized = linalg.init_tensor [%m_size] : tensor<?xi32>
  %zero_matrix = linalg.fill(%c_0, %init_matrix_uninitialized) : i32, tensor<?xi32> -> tensor<?xi32>
  return %zero_matrix : tensor<?xi32>
}

func private @testcase(%matrix : tensor<?x?xi8>) {
  %zeros = util.unfoldable_constant dense<[0, 0, 0]> : tensor<3xi32>
  %zeros_dyn = tensor.cast %zeros : tensor<3xi32> to tensor<?xi32>
  %a = call @reduce(%matrix) : (tensor<?x?xi8>) -> tensor<?xi32>
  check.expect_eq(%a, %zeros_dyn) : tensor<?xi32>
  return
}

func @test() {
  %matrix = util.unfoldable_constant dense<[
      [1, 2, 3, 4],
      [5, 6, 7, 8],
      [9, 10, 11, 12]]> : tensor<3x4xi8>
  %matrix_dyn = tensor.cast %matrix : tensor<3x4xi8> to tensor<?x?xi8>
  call @testcase(%matrix_dyn) : (tensor<?x?xi8>) -> ()
  return
}
```

```
(gdb) bt
#0  0x000000000c886e17 in mlir::detail::IROperandBase::insertInto<mlir::IRObjectWithUseList<mlir::OpOperand> >
    (this=0x6130002d3210, useList=0x0)
    at /usr/local/google/home/benoitjacob/iree/third_party/llvm-project/mlir/include/mlir/IR/UseDefLists.h:82
#1  0x000000000c886d49 in mlir::IROperand<mlir::OpOperand, mlir::Value>::insertIntoCurrent (
    this=0x6130002d3210)
    at /usr/local/google/home/benoitjacob/iree/third_party/llvm-project/mlir/include/mlir/IR/UseDefLists.h:159
#2  0x000000002120b1ad in mlir::IROperand<mlir::OpOperand, mlir::Value>::IROperand (this=0x6130002d3210, 
    owner=0x6130002d3180, value=...)
    at /usr/local/google/home/benoitjacob/iree/third_party/llvm-project/mlir/include/mlir/IR/UseDefLists.h:115
#3  0x00000000212077fb in mlir::OpOperand::IROperand (this=0x6130002d3210)
    at /usr/local/google/home/benoitjacob/iree/third_party/llvm-project/mlir/include/mlir/IR/Value.h:263
#4  0x00000000211f5d87 in mlir::detail::OperandStorage::OperandStorage (this=0x6130002d31c0, 
    owner=0x6130002d3180, trailingOperands=0x6130002d31d0, values=...)
    at /usr/local/google/home/benoitjacob/iree/third_party/llvm-project/mlir/lib/IR/OperationSupport.cpp:227
#5  0x00000000211be072 in mlir::Operation::create (location=..., name=..., resultTypes=..., operands=..., 
    attributes=..., successors=..., numRegions=0)
    at /usr/local/google/home/benoitjacob/iree/third_party/llvm-project/mlir/lib/IR/Operation.cpp:113
#6  0x00000000211bef0f in mlir::Operation::create (location=..., name=..., resultTypes=..., operands=..., 
    attributes=..., successors=..., regions=...)
    at /usr/local/google/home/benoitjacob/iree/third_party/llvm-project/mlir/lib/IR/Operation.cpp:49
#7  0x00000000211be767 in mlir::Operation::create (state=...)
    at /usr/local/google/home/benoitjacob/iree/third_party/llvm-project/mlir/lib/IR/Operation.cpp:38
#8  0x0000000020e00aa1 in mlir::OpBuilder::createOperation (this=0x7fffffff7780, state=...)
    at /usr/local/google/home/benoitjacob/iree/third_party/llvm-project/mlir/lib/IR/Builders.cpp:381
#9  0x0000000014f6b1d8 in mlir::OpBuilder::create<mlir::CallOp, llvm::StringRef, llvm::SmallVector<mlir::Type, 6u>&, llvm::SmallVector<mlir::Value, 6u>&> (this=0x7fffffff7780, location=..., args=..., args=..., args=...)
    at /usr/local/google/home/benoitjacob/iree/third_party/llvm-project/llvm/../mlir/include/mlir/IR/Builders.h:434
#10 0x000000001680aa83 in mlir::iree_compiler::IREE::Stream::(anonymous namespace)::expandCallOp (op=..., 
    indexSet=..., subviewMap=...)
    at /usr/local/google/home/benoitjacob/iree/iree/compiler/Dialect/Stream/Transforms/PropagateSubviews.cpp:395
#11 0x0000000016802503 in mlir::iree_compiler::IREE::Stream::(anonymous namespace)::expandSubviews (
    op=0x6110000e49c0, globalMap=..., indexSet=..., subviewMap=...)
    at /usr/local/google/home/benoitjacob/iree/iree/compiler/Dialect/Stream/Transforms/PropagateSubviews.cpp:490
#12 0x0000000016811b1a in mlir::iree_compiler::IREE::Stream::(anonymous namespace)::expandRegion (region=..., 
    globalMap=..., indexSet=..., subviewMap=...)
    at /usr/local/google/home/benoitjacob/iree/iree/compiler/Dialect/Stream/Transforms/PropagateSubviews.cpp:250
#13 0x000000001680a045 in mlir::iree_compiler::IREE::Stream::(anonymous namespace)::expandFuncOp (op=..., 
    globalMap=..., indexSet=..., subviewMap=...)
```

File: iree/compiler/Dialect/Flow/Transforms/test/expand_tensor_shapes.mlir (hunk 2 above)
Comment (benvanik) [reply iree/compiler/Dialect/Flow/Transforms/test/expand_tensor_shapes.mlir:155]: thought we had a check for that?

File: iree/compiler/Dialect/Flow/Transforms/Passes.cpp (hunk 3 above)
Comment (benvanik) [reply iree/compiler/Dialect/Flow/Transforms/Passes.cpp:138]: thx - fixed

File: iree/compiler/Dialect/Flow/Transforms/Passes.cpp [hunk 4]
```diff
@@ -122,62 +122,65 @@ void buildFlowTransformPassPipeline(OpPassManager &passManager,
...
       .addPredicatedPass(
           !clMmt4dTargetOptions.empty(),
           []() {
-            return createConvertLinalgMatmulToMmt4DPass(clMmt4dTargetOptions);
+            return IREE::Flow::createConvertLinalgMatmulToMmt4DPass(clMmt4dTargetOptions);
           })
       // Pad linalg ops
       .addPredicatedPass(clEnablePaddingLinalgOps, []() {
-        return createPadLinalgOpsToIntegerMultiplePass(clLinalgOpsPaddingSize);
+        return IREE::Flow::createPadLinalgOpsToIntegerMultiplePass(clLinalgOpsPaddingSize);
```
Comment (ScottTodd) [inline iree/compiler/Dialect/Flow/Transforms/Passes.cpp:138]: clang-format wants these lines to wrap

https://github.com/google/iree/runs/5336633750?check_suite_focus=true

**Review by ScottTodd — APPROVED** [review]

**Pull request merged**